### PR TITLE
Fix infinite loop on multiple vouchers

### PR DESCRIPTION
**Pull request by tomlev**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.4.x
| Description?  | fix for infinite loop when multiple vouchers without code are set simultaneously
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Boom 5477
| How to test?  | add 2 percent vouchers with empty code in BO, add a product in FO : cart must have the product and the both vouchers applied

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9036)

### CODE REVIEW

File: tests/Unit/Core/Cart/AbstractCartTest.php [hunk 1]
```diff
@@ -229,60 +234,74 @@ protected function getCartRuleFromFixtureId($id)
...
+            $product = $this->getProductFromFixtureId($cartRuleData['productGiftId']);
+            if ($product === null) {
+                // if product does not exist, skip this rule
+                return;
             }
-            $cartRule->add();
-            $this->cartRules[$k] = $cartRule;
+            $cartRule->gift_product = $product->id;
         }
+        $now = new \DateTime();
```
Comment (PierreRambaud) [inline tests/Unit/Core/Cart/AbstractCartTest.php]: Could be better by using namespace, even for Date classes.

**Review by PierreRambaud — DISMISSED** [review]

File: tests/Unit/Core/Cart/AbstractCartTest.php [hunk 2]
```diff
@@ -229,60 +236,74 @@ protected function getCartRuleFromFixtureId($id)
...
 
+    protected function addCartRuleToCart($cartRuleFixtureId)
+    {
+        $cartRule = $this->getCartRuleFromFixtureId($cartRuleFixtureId);
+        if ($cartRule !== null) {
+            $this->cartRulesInCart[] = $cartRule;
+            $this->cart->addCartRule($cartRule->id);
+        }
+    }
+
```
Comment (PierreRambaud) [inline tests/Unit/Core/Cart/AbstractCartTest.php:308]: Oops, useless return

File: tests/Unit/Core/Cart/Calculation/CartRules/CartRulesWithoutCodeTest.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        $expectedTotal= (1 - $cartRulesData[14]['percent'] / 100)
+                        * (1 - $cartRulesData[15]['percent'] / 100)
+                        * (3 * static::PRODUCT_FIXTURES[1]['price']
+                           + 2 * static::PRODUCT_FIXTURES[2]['price']
+                           + static::PRODUCT_FIXTURES[3]['price'])
+                        + static::DEFAULT_SHIPPING_FEE + static::DEFAULT_WRAPPING_FEE;
+        $this->compareCartTotalTaxIncl($expectedTotal,true);
+    }
+
```
Comment (PierreRambaud) [inline tests/Unit/Core/Cart/Calculation/CartRules/CartRulesWithoutCodeTest.php]: Useless return :)

File: tests/Unit/Core/Cart/Calculation/CartRules/CartRulesWithoutCodeTest.php [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * these tests aim to check the correct calculation of cart total when applying cart rules without code
+ *
+ * products are inserted as fixtures
+ * products are inserted in cart from data providers
+ * cart rules are inserted from data providers
+ */
+class CartRulesWithoutCodeTest extends AbstractCartCalculationTest
+{
+
```
Comment (PierreRambaud) [inline tests/Unit/Core/Cart/Calculation/CartRules/CartRulesWithoutCodeTest.php:40]: Useless line break :)

File: tests/Unit/Core/Cart/AbstractCartTest.php (hunk 2 above)
Comment (tomlev) [reply tests/Unit/Core/Cart/AbstractCartTest.php:308]: we need it to not add a cart rule depending on a not-existing product

Comment (PierreRambaud) [reply tests/Unit/Core/Cart/AbstractCartTest.php:308]: Thinking about ret key, my bad :)

**Review by PierreRambaud — APPROVED** [review]
Will see after for lines break :)

**Comment by mickaelandrieu** [comment]
@tomlev we probably need this great patch to be shipped into 1.7.4.x branch, am I right?

Would you consider to rebase your contribution on `1.7.4.x`?

**Pull request merged**